cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py

```diff
@@ -1351,19 +1363,92 @@ def all_kube_system_pods_running():
 
     try:
         output = check_output(cmd).decode('utf-8')
+        result = json.loads(output)
     except CalledProcessError:
         hookenv.log('failed to get kube-system pod status')
         return False
+    hookenv.log('Checking system pods status: {}'.format(', '.join(
+        '='.join([pod['metadata']['name'], pod['status']['phase']])
+        for pod in result['items'])))
+
+    all_pending = all(pod['status']['phase'] == 'Pending'
+                      for pod in result['items'])
+    if is_state('endpoint.gcp.ready') and all_pending:
+        poke_network_unavailable()
+        return False
 
-    result = json.loads(output)
-    for pod in result['items']:
-        status = pod['status']['phase']
-        # Evicted nodes should re-spawn
-        if status != 'Running' and \
-           pod['status'].get('reason', '') != 'Evicted':
-            return False
+    # All pods must be Running or Evicted (which should re-spawn)
+    all_running = all(pod['status']['phase'] == 'Running' or
+                      pod['status'].get('reason', '') == 'Evicted'
+                      for pod in result['items'])
+    return all_running
 
-    return True
+
+def poke_network_unavailable():
+    """
+    Work around https://github.com/kubernetes/kubernetes/issues/44254 by
+    manually poking the status into the API server to tell the nodes they have
+    a network route.
+
+    This is needed because kubelet sets the NetworkUnavailable flag and expects
+    the network plugin to clear it, which only kubenet does. There is some
+    discussion about refactoring the affected code but nothing has happened
+    in a while.
+    """
+    if is_state('kubernetes-master.networkunavailble.cleared'):
+        return
```

Comment:
If I'm reading this right, the master will clear NetworkUnavailable from all nodes once, then never again.

What about when new kubernetes-worker units are added? The new nodes will need to have the condition cleared too, right?

Reply:
I was under the impression that this issue was caused by a race condition that wouldn't affect new units, but it seems I actually had it backwards.  With the current logic, it's still possible for even the first unit(s) to get past the workaround before it's ready and then end up stuck.  I went with the more conservative approach because I didn't want to accidentally mask a real networking config issue, but I think it's safe to assume that if we have the cloud provider set and we see this particular status, it's incorrect and should be poked.  I'll update.